src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -1005,32 +1006,89 @@ and tc_maybe_toplevel_term env (e:term) : term                  (* type-checked
     check_inner_let_rec env top
 
 and tc_match (env : Env.env) (top : term) : term * lcomp * guard_t =
+
+  (*
+   * AR: Typechecking of match expression:
+   *
+   * match expressions may be optionally annotated with a `return` annotation
+   *   for dependent pattern matching
+   *
+   * When the return annotation is not supplied, we:
+   *   -- typecheck the scrutinee
+   *   -- typecheck the branches with
+   *      -- if the expected type is not set in the env, then create a new uvar for it
+   *      -- a new bv, guard_x below, as the scrutinee expression
+   *   -- combine the computation types of the branches (TcUtil.bind_cases)
+   *      -- with the if_the_else combinator, also adding pattern exhaustiveness checks
+   *   -- bind the scrutinee computation type with the combined branches (using guard_x as the bv in bind)
+   *
+   * When the return annotation is supplied:
+   *   -- typecheck the scrutinee
+   *   -- typecheck the return annotation
+   *   -- typecheck the branches with
+   *      -- env with expected type unset
+   *      -- a new bv, guard_x, as the scrutinee expression
+   *      -- substituting scrutinee bv by guard_x in the return annotation
+   *      -- in tc_eqn: substituting guard_x with the pattern expression, and ascribing it on the branch exp
+   *   -- if the return annotation was a type:
+   *      -- TcUtil.bind_cases as before
+   *   -- if the return annotation was a computation type:
+   *      -- no need to bind cases, take the comp annotation as is
+   *      -- but add the pattern exhaustiveness check in the guard
+   *      -- need to substitute guard_x by scrutinee bv back
+   *   -- bind the scrutinee computation type with the combined branches (using guard_x as the bv in bind)
+   *)
+
   match (SS.compress top).n with
-  | Tm_match(e1, eqns) ->
-    let env1, topt = Env.clear_expected_typ env in
-    let env1 = instantiate_both env1 in
-    let e1, c1, g1 = tc_term env1 e1 in
-    // GM: for some reason, without this annotation on eqns,
-    //     F* fails to do inference. A uvar at type Type gets
-    //     solved to `list u#0`, which is ill-typed.
-    let e1, c1, (eqns : list<S.branch>) =
-        match TcUtil.coerce_views env e1 c1 with
-        | Some (e1, c1) -> (e1, c1, eqns)
-        | None -> (e1, c1, eqns)
-    in
-    let env_branches, res_t, g1 =
-      match topt with
-      | Some t -> env, t, g1
-      | None ->
-        let k, _ = U.type_u() in
-        let res_t, _, g = TcUtil.new_implicit_var "match result" e1.pos env k in
-        Env.set_expected_typ env res_t, res_t, Env.conj_guard g1 g in
+  | Tm_match(e1, ret_opt, eqns) ->  //ret_opt is the return annotation
+    let e1, c1, g1 = tc_term
+      ({(env |> Env.clear_expected_typ |> fst) with instantiate_imp = true})
```

Comment:
May be a bit more abstract to write it a

```
env |> Env.clear_expected_typ |> fst |> instantiate_both
```

?

Reply:
Done, thanks!